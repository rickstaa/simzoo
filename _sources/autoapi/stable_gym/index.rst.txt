stable_gym
==========

.. py:module:: stable_gym

.. autoapi-nested-parse::

   Entry_point that register the Stable Gym gymnasium environments.



Subpackages
-----------

.. toctree::
   :maxdepth: 1

   /autoapi/stable_gym/common/index
   /autoapi/stable_gym/envs/index


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/stable_gym/version/index


Attributes
----------

.. autoapisummary::

   stable_gym.__version__
   stable_gym.__version_tuple__
   stable_gym.ENVS


Classes
-------

.. autoapisummary::

   stable_gym.MaxEpisodeStepsInjectionWrapper


Package Contents
----------------

.. py:class:: MaxEpisodeStepsInjectionWrapper(env)

   Bases: :py:obj:`gymnasium.Wrapper`


   A gymnasium wrapper that injects the ``max_episode_steps`` attribute of the
   ``TimeLimit`` wrapper into the base environment as the
   ``_time_limit_max_episode_steps`` attribute. If the environment is not wrapped in
   a ``TimeLimit`` wrapper, then the ``_time_limit_max_episode_steps`` attribute is
   set to ``None``.

   Wrap a gymnasium environment.
   :param env: The gymnasium environment.
   :type env: gym.Env


   .. py:attribute:: max_episode_steps


.. py:data:: __version__
   :value: '0.20.7'


.. py:data:: __version_tuple__

.. py:data:: ENVS

