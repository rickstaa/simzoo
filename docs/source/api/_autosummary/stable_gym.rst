stable\_gym
===========

.. automodule:: stable_gym
  
   
   
   

   
   
   

   
   
   

   
   
   



.. rubric:: Modules

.. autosummary::
   :toctree:
   :template: custom-module-template.rst
   :recursive:

   stable_gym.common
   stable_gym.envs

