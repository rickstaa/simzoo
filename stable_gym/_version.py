# Stores the package version number so that it can be accessed from other modules.
version = "0.6.4"
version_tuple = version.split(".")
