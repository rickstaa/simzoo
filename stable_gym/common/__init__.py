"""Contains several functions that can be used across all the Stable Gym gymnasium
environments.
"""
from stable_gym.common.disturber import Disturber  # noqa: F401
