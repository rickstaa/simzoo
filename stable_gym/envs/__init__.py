"""The Stable Gym gymnasium environments."""
