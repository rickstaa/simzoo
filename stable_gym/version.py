# Stores the package version number so that it can be accessed from other modules.
__version__ = "0.6.38"
__version_tuple__ = __version__.split(".")
