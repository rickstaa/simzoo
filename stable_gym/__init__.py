"""Module that register the Stable Gym gymnasium environments."""
import gymnasium as gym
from gymnasium.envs.registration import register

# Make module version available.
from .version import __version__, __version_tuple__

# Available environments.
# TODO: Update reward thresholds.
ENVS = {
    "Oscillator-v1": {
        "module": "stable_gym.envs.biological.oscillator.oscillator:Oscillator",
        "max_step": 400,
        "reward_threshold": 300,
    },
    "OscillatorComplicated-v1": {
        "module": "stable_gym.envs.biological.oscillator_complicated.oscillator_complicated:OscillatorComplicated",
        "max_step": 400,
        "reward_threshold": 300,
    },
    "CartPoleCost-v1": {
        "module": "stable_gym.envs.classic_control.cartpole_cost.cartpole_cost:CartPoleCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "Ex3EKF-v1": {
        "module": "stable_gym.envs.classic_control.ex3_ekf.ex3_ekf:Ex3EKF",
        "max_step": 400,
        "reward_threshold": 300,
    },
    "AntCost-v1": {
        "module": "stable_gym.envs.mujoco.ant_cost.ant_cost:AntCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "HalfCheetahCost-v1": {
        "module": "stable_gym.envs.mujoco.half_cheetah_cost.half_cheetah_cost:HalfCheetahCost",
        "max_step": 200,
        "reward_threshold": 300,
    },
    "HopperCost-v1": {
        "module": "stable_gym.envs.mujoco.hopper_cost.hopper_cost:HopperCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "SwimmerCost-v1": {
        "module": "stable_gym.envs.mujoco.swimmer_cost.swimmer_cost:SwimmerCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
    "Walker2dCost-v1": {
        "module": "stable_gym.envs.mujoco.walker2d_cost.walker2d_cost:Walker2dCost",
        "max_step": 250,
        "reward_threshold": 300,
    },
}

for env, val in ENVS.items():
    if env not in gym.envs.registry:  # NOTE: Required because we use namespace packages
        register(
            id=env,
            entry_point=val["module"],
            max_episode_steps=val["max_step"],
            reward_threshold=val["reward_threshold"],
        )
