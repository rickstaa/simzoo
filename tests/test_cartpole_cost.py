"""Test if the CartPoleCost environment still behaves like the original CartPole
environment when the same environment parameters are used.
"""
import math
import random

import gymnasium as gym
import numpy as np
from gymnasium.logger import ERROR

import pytest
import stable_gym  # noqa: F401

gym.logger.set_level(ERROR)


class TestCartPoleCostEqual:
    @pytest.fixture
    def env_original(self):
        """Create original CartPole environment."""
        return gym.make("CartPole")

    @pytest.fixture
    def env_cost_aligned(self):
        """Create CartPoleCost environment and align the parameters with the
        original CartPole environment.
        """
        env_cost = gym.make(
            "CartPoleCost",
        )

        # Overwrite changed parameters.
        env_cost.unwrapped.force_mag = 10.0
        env_cost.unwrapped.theta_threshold_radians = 12 * 2 * math.pi / 360
        env_cost.unwrapped.x_threshold = 2.4
        env_cost.unwrapped.max_v = np.finfo(np.float64).max
        env_cost.unwrapped.max_x = np.finfo(np.float64).max
        env_cost.unwrapped._init_state_range = {
            "low": [-0.05, -0.05, -0.05, -0.05],
            "high": [0.05, 0.05, 0.05, 0.05],
        }
        return env_cost

    @pytest.fixture
    def env_cost(self):
        """Create CartPoleCost environment."""
        return gym.make(
            "CartPoleCost",
        )

    def test_equal_reset(self, env_original, env_cost_aligned):
        """Test if reset behaves the same."""
        # Perform reset and check if observations are equal.
        observation, _ = env_original.reset(seed=42)
        observation_cost, _ = env_cost_aligned.reset(seed=42)
        assert np.allclose(
            observation, observation_cost
        ), f"{observation} != {observation_cost}"

    def test_equal_steps(self, env_original, env_cost_aligned):
        """Test if steps behave the same."""
        # Perform several steps and check if observations are equal.
        env_original.reset(seed=42), env_cost_aligned.reset(seed=42)
        for _ in range(10):
            discrete_action = random.randint(0, 1)
            continuous_action = (
                np.array([-10]) if discrete_action == 0 else np.array([10])
            )
            observation, _, _, _, _ = env_original.step(discrete_action)
            observation_cost, _, _, _, _ = env_cost_aligned.step(continuous_action)
            assert np.allclose(
                observation, observation_cost
            ), f"{observation} != {observation_cost}"

    def test_snapshot(self, snapshot, env_cost):
        """Test if the 'CartPoleCost' environment is still equal to snapshot."""
        observation, info = env_cost.reset(seed=42)
        assert (observation == snapshot).all()
        assert info == snapshot
        env_cost.action_space.seed(42)
        for _ in range(5):
            action = env_cost.action_space.sample()
            observation, reward, terminated, truncated, info = env_cost.step(action)
            assert (observation == snapshot).all()
            assert reward == snapshot
            assert terminated == snapshot
            assert truncated == snapshot
            assert info == snapshot
