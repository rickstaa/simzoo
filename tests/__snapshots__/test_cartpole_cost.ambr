# serializer version: 1
# name: TestCartPoleCostEqual.test_snapshot
  0.10958241942238534
# ---
# name: TestCartPoleCostEqual.test_snapshot.1
  -0.024448624099179084
# ---
# name: TestCartPoleCostEqual.test_snapshot.10
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.11
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.12
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.13
  0.11283397455824112
# ---
# name: TestCartPoleCostEqual.test_snapshot.14
  0.1373521677359711
# ---
# name: TestCartPoleCostEqual.test_snapshot.15
  0.14115849335141145
# ---
# name: TestCartPoleCostEqual.test_snapshot.16
  -0.07676570293056856
# ---
# name: TestCartPoleCostEqual.test_snapshot.17
  3.270741362539796
# ---
# name: TestCartPoleCostEqual.test_snapshot.18
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.19
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.2
  0.14343916796455297
# ---
# name: TestCartPoleCostEqual.test_snapshot.20
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.21
  0.11558101791296055
# ---
# name: TestCartPoleCostEqual.test_snapshot.22
  0.41483476162456334
# ---
# name: TestCartPoleCostEqual.test_snapshot.23
  0.13962317929280008
# ---
# name: TestCartPoleCostEqual.test_snapshot.24
  -0.44748678167775646
# ---
# name: TestCartPoleCostEqual.test_snapshot.25
  3.199988707897624
# ---
# name: TestCartPoleCostEqual.test_snapshot.26
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.27
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.28
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.29
  0.12387771314545182
# ---
# name: TestCartPoleCostEqual.test_snapshot.3
  0.07894721162374557
# ---
# name: TestCartPoleCostEqual.test_snapshot.30
  0.5667135704060806
# ---
# name: TestCartPoleCostEqual.test_snapshot.31
  0.13067344365924494
# ---
# name: TestCartPoleCostEqual.test_snapshot.32
  -0.6321720124467993
# ---
# name: TestCartPoleCostEqual.test_snapshot.33
  2.8029394711714257
# ---
# name: TestCartPoleCostEqual.test_snapshot.34
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.35
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.36
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.37
  0.13521198455357342
# ---
# name: TestCartPoleCostEqual.test_snapshot.38
  0.24856720024803347
# ---
# name: TestCartPoleCostEqual.test_snapshot.39
  0.11803000341030895
# ---
# name: TestCartPoleCostEqual.test_snapshot.4
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.40
  -0.12071230469371952
# ---
# name: TestCartPoleCostEqual.test_snapshot.41
  2.286834996999153
# ---
# name: TestCartPoleCostEqual.test_snapshot.42
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.43
  False
# ---
# name: TestCartPoleCostEqual.test_snapshot.44
  dict({
  })
# ---
# name: TestCartPoleCostEqual.test_snapshot.5
  0.10909344694040175
# ---
# name: TestCartPoleCostEqual.test_snapshot.6
  0.18702638089196827
# ---
# name: TestCartPoleCostEqual.test_snapshot.7
  0.14501811219702787
# ---
# name: TestCartPoleCostEqual.test_snapshot.8
  -0.19298094228082074
# ---
# name: TestCartPoleCostEqual.test_snapshot.9
  3.4520313962115345
# ---
